### PR TITLE
Create code coverage build.

### LINKED ISSUE #11: Track code coverage metrics
While code coverage is an imperfect metric of test quality (necessary, but not sufficient), it's an additional useful signal to keep in mind when assessing the quality of the overall test suite.

There are several SaaS options which are free for open-source projects:

* [Codecov](https://codecov.io/)
* [Coveralls](https://coveralls.io/)
* [Circle CI](https://circleci.com/docs/code-coverage/) (code coverage + CI in one)

I've heard good things about Codecov, and we're using it in [another project](https://github.com/JanusGraph/janusgraph) and it works well, and is easy to set up.

Here are the [C++ setup instructions for Codecov](https://github.com/codecov/example-c), which seem quite simple and straight-forward.

As an aside: Codecov uses [gcov](https://gcc.gnu.org/onlinedocs/gcc/Gcov.html) for C++ coverage (via Clang's `-coverage` flag), which is also what the [Google blog post on code coverage from 2014](https://testing.googleblog.com/2014/07/measuring-coverage-at-google.html) talks about, so seems like this is a good choice.

### PR DESCRIPTION
**Pull request by coryan**
This fixes #11.

### CODE REVIEW

File: ci/upload-codecov.sh [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+         "without further action."
+    exit 0
+fi
+
+if [ "${BUILD_TYPE:-Release}" != "Coverage" ]; then
+    echo "Code coverage not enabled, exit successfully without further action."
+    exit 0
+fi
+
+if [ -z "${CODECOV_TOKEN+x}" ]; then
```
Comment (mbrukman) [inline ci/upload-codecov.sh]: The `+x` here is extra, no? If you want to test for empty, just test for empty.

File: README.md [hunk 2]
```diff
@@ -1,10 +1,14 @@
 # Google Cloud Platform C++ Client Libraries
 
 [![Travis CI status][travis-shield]][travis-link]
+[![Codecov Coverage status][codecov-shield]][codecov-link]
 
 [travis-shield]: https://travis-ci.org/GoogleCloudPlatform/google-cloud-cpp.svg
 [travis-link]: https://travis-ci.org/GoogleCloudPlatform/google-cloud-cpp/builds
```
Comment (mbrukman) [inline README.md]: Drop this blank line?

File: bigtable/CMakeLists.txt [hunk 3]
```diff
@@ -35,6 +35,46 @@ if(NOT MSVC)
   endif()
 endif()
 
+# For the gcc and clang compiler families enable a Coverage build type.
```
Comment (mbrukman) [inline bigtable/CMakeLists.txt]: For the GCC and Clang compiler families, enable a Coverage build type.

File: ci/upload-codecov.sh [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -eu
+
+if [ "${TRAVIS_PULL_REQUEST}" != "false" ]; then
+    echo "Code coverage disabled in pull requests, exit successfully" \
```
Comment (mbrukman) [inline ci/upload-codecov.sh]: Skipping code coverage as it is disabled for pull requests.

File: ci/upload-codecov.sh [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+         "without further action."
+    exit 0
+fi
+
+if [ "${BUILD_TYPE:-Release}" != "Coverage" ]; then
+    echo "Code coverage not enabled, exit successfully without further action."
```
Comment (mbrukman) [inline ci/upload-codecov.sh]: Skipping code coverage as it is disabled for this build.

File: ci/upload-codecov.sh [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    exit 0
+fi
+
+if [ -z "${CODECOV_TOKEN+x}" ]; then
+    echo "CODECOV_TOKEN not configured in Coverage build, exit with error."
+    exit 1
+fi
+
+# Copy the build results out of the docker image
+IMAGE="cached-${DISTRO}-${DISTRO_VERSION}";
```
Comment (mbrukman) [inline ci/upload-codecov.sh]: `readonly`

Also, trailing semicolon is unnecessary on this and next line.

File: ci/upload-codecov.sh [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+sudo docker run --volume $PWD:/d --rm -it ${IMAGE}:tip cp -r /var/tmp/build/gccpp/build-output /d;
```
Comment (mbrukman) [inline ci/upload-codecov.sh]: `"${PWD}:/d"`

`"${IMAGE}:tip"`

File: ci/upload-codecov.sh (hunk 1 above)
Comment (mbrukman) [inline ci/upload-codecov.sh]: Also, how does this token actually propagate to the build? Don't you need an encrypted env var in `.travis.yml` for this to work?

File: ci/upload-codecov.sh [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+# Upload the results using the script from codecov.io
+bash <(curl -s https://codecov.io/bash) >codecov.log 2>&1 \
```
Comment (mbrukman) [inline ci/upload-codecov.sh]: Why not just run `bash <(curl -s https://codecov.io/bash)`? Does it really have > 2K LOC output that we need the added complexity of saving to the log and the `head` / `tail` commands?

File: .codecov.yml [hunk 9]
```diff
@@ -0,0 +1,6 @@
+ignore:
```
Comment (mbrukman) [inline .codecov.yml]: Here are some additional config options for a quieter experience with Codecov:

```yaml
# Don't post a comment on pull requests.
comment: off

# Disable commit statuses
coverage:
  status:
    project: no
    patch: no
    changes: no
```

via https://github.com/JanusGraph/janusgraph/blob/master/codecov.yml

File: README.md (hunk 2 above)
Comment (coryan) [reply README.md]: Done.

File: bigtable/CMakeLists.txt (hunk 3 above)
Comment (coryan) [reply bigtable/CMakeLists.txt]: Done.

File: ci/upload-codecov.sh (hunk 5 above)
Comment (coryan) [reply ci/upload-codecov.sh]: Done.

File: ci/upload-codecov.sh (hunk 1 above)
Comment (coryan) [reply ci/upload-codecov.sh]: If CODECOV_TOKEN is not set then the expansion fails (due to the `set -eu`).  We either remove the `-u` and use `${VAR?}` everywhere else, or use `${VAR+x}` here.

Comment (coryan) [reply ci/upload-codecov.sh]: It does not propagate to the build.  We build with `--coverage` inside docker, then get the files out where `CODECOV_TOKEN` is set.  I wanted to run the upload inside the Docker container, but ran into too many problems.

File: ci/upload-codecov.sh (hunk 6 above)
Comment (coryan) [reply ci/upload-codecov.sh]: Done.

File: ci/upload-codecov.sh (hunk 7 above)
Comment (coryan) [reply ci/upload-codecov.sh]: Done.

File: ci/upload-codecov.sh (hunk 8 above)
Comment (coryan) [reply ci/upload-codecov.sh]: The output is more than 4MB which is the limit for Travis.  The build breaks if you dump the full log.

File: .codecov.yml (hunk 9 above)
Comment (coryan) [reply .codecov.yml]: Thanks!

**Comment by coryan** [comment]
Thanks for the review.  PTAL.

File: ci/upload-codecov.sh [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -eu
+
+if [ "${TRAVIS_PULL_REQUEST}" != "false" ]; then
+    echo "Skipping code coverage as it is disabled for pull requests," \
+         "exit successfully without further action."
```
Comment (mbrukman) [inline ci/upload-codecov.sh]: Drop this line? Exiting is just this script, it doesn't stop the entire build.

File: ci/upload-codecov.sh [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    exit 0
+fi
+
+if [ "${BUILD_TYPE:-Release}" != "Coverage" ]; then
+    echo "Skipping code corage as it is disabled for this build," \
+         "exit successfully without further action."
```
Comment (mbrukman) [inline ci/upload-codecov.sh]: Drop this line? Exiting is just this script, it doesn't stop the entire build.

File: ci/upload-codecov.sh [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+set -eu
+
+if [ "${TRAVIS_PULL_REQUEST}" != "false" ]; then
+    echo "Skipping code coverage as it is disabled for pull requests," \
+         "exit successfully without further action."
+    exit 0
+fi
+
+if [ "${BUILD_TYPE:-Release}" != "Coverage" ]; then
+    echo "Skipping code corage as it is disabled for this build," \
```
Comment (mbrukman) [inline ci/upload-codecov.sh]: s/corage/coverage/

File: ci/upload-codecov.sh (hunk 1 above)
Comment (mbrukman) [reply ci/upload-codecov.sh]: If you want a nice message when `VAR` is empty, don't you want to use `"${VAR:-}"`?

`${VAR+x}` returns:

* `$VAR` iff `$VAR` is set
* empty if `$VAR` is empty
* `x` iff `$VAR` is unset

which is not exactly what you want: you want the last two cases to both be caught by `-z` which means you want to use `${VAR:+x}` but `${VAR:-}` is much simpler since you just care about empty and don't need the "x", so just use that.

Look at the output of this script:

```bash
#!/bin/bash

readonly EMPTY=""
readonly SET=y

echo "unset: ${UNSET+x}"
echo "empty: ${EMPTY+x}"
echo "  set: ${SET+x}"

echo "unset: ${UNSET:+x}"
echo "empty: ${EMPTY:+x}"
echo "  set: ${SET:+x}"
```

File: ci/upload-codecov.sh (hunk 10 above)
Comment (coryan) [reply ci/upload-codecov.sh]: Done.

File: ci/upload-codecov.sh (hunk 12 above)
Comment (coryan) [reply ci/upload-codecov.sh]: Done.

File: ci/upload-codecov.sh (hunk 11 above)
Comment (coryan) [reply ci/upload-codecov.sh]: Done.

**Comment by coryan** [comment]
PTAL.

**Comment by mbrukman** [comment]
I'll merge the PR once the build/test is green.
